packages/plugin-ext/doc/how-to-add-new-plugin-namespace.md

```diff
@@ -90,23 +122,153 @@ function findPlugin(filePath: string): Plugin | undefined {
+## Implement your API object
+
+We create a dedicated API object for each individual plugin as part of the module loading process.
+Each API object is returned as part of the module loading process if a script imports `@bar/foo` and should therefore match the API definition that we provided in the `*.d.ts` file.
+Multiple imports will not lead to the creation of multiple API objects as we cache it in our custom `overrideInternalLoad` function.
 
-Example:
+Example `foo-init.ts` (continued):
 
 ```typescript
-import * as fooApi from '@bar/foo';
 export function createAPIFactory(rpc: RPCProtocol): ApiFactory {
-    const fooBarImpl = new FooBarImpl(rpc);
-    return function (plugin: Plugin): typeof fooApi {
-        const FooBar: typeof fooApi.fooBar = {
-            getFoo(): fooApi.Foo{
-                return fooBarImpl.getFooImpl();
+    const fooExtImpl = new FooExtImpl(rpc);
+    return function (plugin: Plugin): typeof fooBarAPI {
+        const FooBar: typeof fooBarAPI.fooBar = {
+            getFoo(): Promise<fooBarAPI.Foo> {
+                return fooExtImpl.getFooImpl();
             }
         }
-        return <typeof fooApi>{
+        return <typeof fooBarAPI>{
             fooBar : FooBar
         };
     }
+}
+```
+
+In the example above the API object creates a local object that will fulfill the API contract.
+The implementation details are hidden by the object and it could be a local implementation that only lives inside the plugin host but it could also be an implementation that uses the `RPCProtocol` to communicate with the main application to retrieve additional information.
+
+### Implementing Main-Ext communication
+
+In this document, we will only highlight the individual parts needed to establish the communication between the main application and the external plugin host.
+For a more elaborate example of an API that communicates with the main application, please have a look at the definition of the [Theia Plugin API](https://github.com/eclipse-theia/theia/blob/master/doc/Plugin-API.md).
+
+First, we need to establish the communication on the RPC protocol by providing an implementation for our own side and generating a proxy for the opposite side.
+Proxies are identified using dedicated identifiers so we set them up first, together with the expected interfaces.
+`Ext` and `Main` interfaces contain the functions called over RCP and must start with `$`.
+Due to the asynchronous nature of the communication over RPC, the result should always be a `Promise` or `PromiseLike`.
+
+Example `common/foo-api-rpc.ts`:
+
+```typescript
+export interface FooMain {
+    $getFooImpl(): Promise<Foo>;
+}
+
+export interface FooExt {
+    // placeholder for callbacks for the main application to the extension
+}
+
+// Plugin host will obtain a proxy using these IDs, main application will register an implementation for it.
+export const PLUGIN_RPC_CONTEXT = {
+    FOO_MAIN: <ProxyIdentifier<FooMain>>createProxyIdentifier<FooMain>('FooMain')
```

Comment:
Very nice job on this new section!

If I'm not mistaken, I think the `<ProxyIdentifier<FooMain>>` can be omitted.  Same idea on line 180.

Reply:
Absolutely correct, thanks!